nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyArrayTests.java

```diff
@@ -0,0 +1,85 @@
+package com.uber.nullaway;
+
+import com.google.errorprone.CompilationTestHelper;
+import java.util.Arrays;
+import org.junit.Test;
+
+public class NullAwayJSpecifyArrayTests extends NullAwayTestsBase {
+
+  @Test
+  public void arrayTopLevelAnnotationDereference() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  static Integer @Nullable [] fizz = {1};",
+            "  static void foo() {",
+            "     // BUG: Diagnostic contains: dereferenced expression fizz is @Nullable",
+            "     int bar = fizz.length;",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void arrayTopLevelAnnotationAssignment() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  Object foo = new Object();",
+            "  void m( Integer @Nullable [] bar) {",
+            "      // BUG: Diagnostic contains: assigning @Nullable expression to @NonNull field",
+            "      foo = bar;",
+            "  }",
+            "}")
+        .doTest();
+  }
+
+  @Test
+  public void arrayContentsAnnotationDereference() {
+    makeHelper()
+        .addSourceLines(
+            "Test.java",
+            "package com.uber;",
+            "import org.jspecify.annotations.Nullable;",
+            "class Test {",
+            "  static @Nullable String [] fizz = {\"1\"};",
+            "  static void foo() {",
+            "      // TODO: This should report an error due to dereference of @Nullable fizz[0]",
```

Comment:
For the first one, I think we need to edit the `ARRAY_ACCESS` switch case [here](https://github.com/uber/NullAway/blob/97e92b9d32603ccb9bc7bc610e2d707751d4ca2b/nullaway/src/main/java/com/uber/nullaway/NullAway.java#L2255).
For the second one, do you mean to ask if this test case is redundant because `arrayContentsAnnotationAssignment`
already [handles](https://github.com/uber/NullAway/blob/5726080dda133769c31edd2a7d872fe67e258d39/nullaway/src/test/java/com/uber/nullaway/NullAwayJSpecifyArrayTests.java#L72) the dereference case?

Reply:
For the second I am asking why we catch `array.length` but not `array[i]`, since those both are handled by `matchDereference` in NullAway.java. Such a direct dereference seems dataflow independent, but I didn't have the chance to dig super deeply right now, so maybe I am wrong.

Reply:
For this expression `fizz[0].length()`, we have the declaration `@Nullable String [] fizz`.  So, `fizz` is `@NonNull`, but `fizz[0]` is `@Nullable`.  If `fizz` were `@Nullable` we would detect that, same as `fizz.length`.  Should we add a test for `fizz[0]` where `fizz` is `@Nullable`?  As you noted they go through the same code path.

Reply:
I think from understood and worked on, we catch both right now when annotation is on the top-level type (`String @Nullable [] array`) since both call `matchDereference` as you rightly mentioned. However, the _**eventual**_ goal of this PR is to make the distinction between the two when annotation is on contents (`@Nullable String [] array`), and only catch `array[i]` in the latter case. Am i correct? @msridhar

Reply:
@armughan11 you are correct about the eventual goal, but it's not part of the goal for _this_ PR.  We try to keep PRs small, and this one adds the functionality of checking for `@Nullable` on a top-level type in the right place according to JSpecify, when in JSpecify mode.  We can handle `@Nullable` on array element types in follow-up PRs.

Reply:
Right. I am too used to the old (incorrect) semantics of `@Nullable Foo[]`, my bad...

> Should we add a test for fizz[0] where fizz is @Nullable? As you noted they go through the same code path.

Yes. Having this test case seems wise, it being the same code path is an implementation detail, so good to test it separately. It also makes the current test case clearer by comparison!

Reply:
5816c99f58dfb9ea8e8452b61977c3268836382e